### ISSUE TITLE
Content plays before ad (Reopening #167)

### ISSUE THREAD

**Issue by wilbert-abreu** [googleads/videojs-ima#558]
@shawnbuso I'm using version `1.2.1`and still seeing the content play briefly before the ad and I've tried all the recommended solutions.

Please let me know if I should rollback to a specific version. Thanks!

**Comment by shawnbuso** [comment]
Hey there,

If you're calling videoPlayer.play() in your implementation, that can be removed as of IMA plugin v1.0. If that's not the issue, could you send me a sample page to take a look at?

Thanks!
Shawn

**Comment by mysuf** [comment]
I can confirm this behaviour. With some connection issues it can be even few seconds when preroll starts - and all that time, video is playing default content. It could be related to what I said in #539 (before play and before ended impl)

**Comment by shawnbuso** [comment]
Oh interesting - I am able to repro this if I throttle my network connection. This might be a contrib-ads issue - I'm seeing the following state transitions when I load my sample page with a bad network connection:

```
VIDEOJS: ADS: Received adsready event (BeforePreroll)
VIDEOJS: ADS: Received play event (BeforePreroll)
VIDEOJS: ADS: t -> n
VIDEOJS: ADS: Triggered readyforpreroll event (Preroll)
VIDEOJS: ADS: adtimeout (Preroll)
VIDEOJS: ADS: n -> t
VIDEOJS: ADS: t -> t
VIDEOJS: ADS: Starting ad break
VIDEOJS: ADS: Ending ad break
VIDEOJS: ADS: t -> t
```
So I tell it ads are ready, it fires ready for preorll, then immeduately signals adtimeout. But after adtimeout, the pre-roll starts. This happens regardless of what I set for prerollTimeout in contrib-ads.

@incompl, does this look like something going wrong in our implementation, or something that needs to be addressed from within contrib-ads?

**Comment by incompl** [comment]
Can you share the sample page so I can take a look? Also, you can fix the state logging by using unminified code (will be fixed in next contrib-ads release so it will minify correctly)

**Comment by shawnbuso** [comment]
I asked a question without providing a sample page... I have become what I hate.

[Sample page here](https://shawnbuso.github.io/videojs-ima/examples/simple/), though I can only repro if I use the Chrome network tab to throttle the connection to "Slow 3G," which makes this painful.

I also see different results based on whether or not I wait for "adsready" to fire before clicking the play button. If I click play *after* seeing adsready, I get this:
```
VIDEOJS: ADS: Received adsready event (BeforePreroll)
VIDEOJS: ADS: Received play event (BeforePreroll)
VIDEOJS: ADS: BeforePreroll -> Preroll
VIDEOJS: ADS: Triggered readyforpreroll event (Preroll)
VIDEOJS: ADS: adtimeout (Preroll)
VIDEOJS: ADS: Preroll -> ContentPlayback
VIDEOJS: ADS: ContentPlayback -> Midroll
VIDEOJS: ADS: Starting ad break
```
If I click it *before* seeing adsready, I get this:
```
VIDEOJS: ADS: Received play event (BeforePreroll)
VIDEOJS: ADS: BeforePreroll -> Preroll
VIDEOJS: ADS: adtimeout (Preroll)
VIDEOJS: ADS: Preroll -> ContentPlayback
VIDEOJS: ADS: Received adsready event (ContentPlayback)
VIDEOJS: ADS: Triggered readyforpreroll event (ContentPlayback)
VIDEOJS: ADS: ContentPlayback -> Midroll
VIDEOJS: ADS: Starting ad break
```

**Comment by incompl** [comment]
I'm getting a 404 for `https://shawnbuso.github.io/videojs-ima/node_modules/videojs-contrib-ads/dist/videojs.ads.js` on the sample page.

**Comment by shawnbuso** [comment]
Whoops - rogue gitignore. I added it but it looks like it's still working its way through the gh-pages cache.

**Comment by incompl** [comment]
From the logs and example, it looks like the ad is playing even though an ad timeout happened. We refer to this as a _soft timeout_. Here is a piece of information from the [contrib-ads documentation](https://github.com/videojs/videojs-contrib-ads/#timeout) that may be the solution to your problem:

> Some ad plugins may want to play a preroll ad even after the timeout has expired and content has begun playing. To facilitate this, videojs-contrib-ads will respond to an adsready event during content playback with a readyforpreroll event. If you want to avoid this behavior, make sure your plugin does not send adsready if player.ads.isInAdMode() is false.

**Comment by shawnbuso** [comment]
Aha that makes sense, I can create a PR to check to make sure we're not firing adsready if we're not in ad mode. Also I was in error when I said this was happening regardless of the prerollTimeout setting - I had some code that was overwriting the custom timeout I was trying to provide. I'll get that fixed too.

**Comment by shawnbuso** [comment]
Actually, it looks a bit more complicated than that - it looks like my current implementation, with a fast network, is firing adsready before any contrib-ads state transitions happen, and isInAdMode() is returning false. It seems like maybe I need to check that player.ads.isInAdMode() is not false *and* that we've started playback. Does that make sense? Or should I not be triggering adsready at all until play has been fired?

**Comment by incompl** [comment]
Oh I see, I hadn't thought of that. Checking that the player is not in ad mode and that content is playing seems reasonable to me. It is OK to trigger adsready before ad mode.

**Comment by shawnbuso** [comment]
OK I'm very close to a solution - the only problem I have remaining is in the case where I fire adsReady before play has been clicked, and then loading the pre-roll ends up timing out[1]. When that happens, my pre-roll ad eventually loads and ends up triggering ad mode and playing as a mid-roll, like this:
```
VIDEOJS: ADS: Received adsready event (BeforePreroll)
VIDEOJS: ADS: Received play event (BeforePreroll)
VIDEOJS: ADS: BeforePreroll -> Preroll
VIDEOJS: ADS: Triggered readyforpreroll event (Preroll)
VIDEOJS: ADS: adtimeout (Preroll)
VIDEOJS: ADS: Preroll -> ContentPlayback
VIDEOJS: ADS: ContentPlayback -> Midroll
VIDEOJS: ADS: Starting ad break
```
Is there any way for me to get notified when the contrib-ads plugin reaches "adtimeout (Preroll)" so I can set a flag to skip the pre-roll when it eventually loads?

[1] I know ideally I shouldn't fire adsready until the pre-roll is ready, however with the IMA SDK we're loading a playlist of ads, firing adsready, and then on readyforpreroll we get the pre-roll from the playlist and play it. Unfortunately "get the preroll from the playlist and play it" is all one step for the IMA SDK - I can't get the preroll from the playlist, fire adsready, and then play it later.

**Comment by incompl** [comment]
For this you'll want to listen to the `adtimeout` event:

```js
player.on('adtimeout', () => {
    // do what needs doing
});
```

**Comment by shawnbuso** [comment]
Ah I didn't realize that was fired, thought it was internal to contrib-ads.

That's working, but now I'm hitting other issues internal to the plugin. If I skip calling adsready I never set up the triggers in IMA that cause mid- and post-rolls to play, so what I have now works but if the pre-roll times out it never tries to play mid-roll or post-roll ads either, which isn't ideal. Fixing this bug might end up turning into a small refactor of the plugin. I'm going to shelve it for the weekend and see if I have any moments of sudden clarity before Monday. Thanks for all the help!

**Comment by mysuf** [comment]
Maybe its not so bad idea .  When I tried to understand your source code, I found there so many layers where each refers to another and almost all has duplicated methods (like .onSomething). When I wanted to expose some internal event, it was spaghetti code. Is that all necessary?

**Comment by shawnbuso** [comment]
We did a refactor a few months back and redesigned it as best we could - the issue is that we have multiple components that all rely on each other, so it's hard to avoid spaghettification. For example the ad UI needs to call methods of the IMA SDK implementation, but that implementation also has to call methods of the ad ui, and both need to call methods of the player wrapper, which needs to call methods of both of those. What we have now is some spaghetti, but it's better than what we had before which was just one massive file with no division of responsibility. I'm going to keep looking into the best way to handle this but it might take a bit to arrive at a nice solution, especially considering that right now we can only reproduce on very slow connections, so it's not affecting all users.

**Comment by adripanico** [comment]
Are these changes still applicable? I'm suffering this issue but I'm not sure if they still have sense after all the new releases for videojs and videojs-contrib-ads.

**Comment by Kiro705** [comment]
Hi @adripanico,

Thanks for letting us know. Would it be possible for you to explain how to reproduce the scenario related to the content playing before the preroll?

It has been a long time since there was any mention of this issue and I want to confirm its the same thing.

Thanks,
Jackson
IMA SDK DevRel

**Comment by ckybonist** [comment]
I'm also encounter similar issue. My situation is kind of weird:  
1. When `adTagUrl` is the [ima-sdk sample url](https://developers.google.com/interactive-media-ads/docs/sdks/html5/tags?hl=zh-TW), then the ads is shown as expected, however
2. if I change the `adTagUrl` to my custom url (which is a private ad-server vmap api), turns out there is no adbreaks appear in my video. 

Here is my log:
```
VIDEOJS: ADS: Received play event (BeforePreroll)
VIDEOJS: ADS: BeforePreroll -> Preroll
VIDEOJS: ADS: Received adsready event (Preroll)
VIDEOJS: ADS: Triggered readyforpreroll event (Preroll)
VIDEOJS: ADS: adtimeout (Preroll)
VIDEOJS: ADS: Preroll -> ContentPlayback
```

My code is basically same as the `videojs-ima/examples/simple`. And I've tested my vmap url in the [official demo page](http://googleads.github.io/videojs-ima/examples/advanced/) did correctly render adbreaks.

Help !!!

**Comment by Kiro705** [comment]
I'm going to close this issue, but feel free to reopen if it still occurs.

**Comment by adripanico** [comment]
Sorry! I read the message on the phone and I totally forgot to answer it.

Actually, it is going to be too difficult to reproduce my solution since I have created a bundle with video.js, some plugins and some other customization.

In any case, it seems we have finally avoided the problem with two approaches. First, adding the following to the IMA plugin initialization:

```
contribAdsSettings: {
   prerollTimeout: 5000,
},
```

and second, adding the following style:

```
.vjs-ad-loading > video, .vjs-ad-loading > .vjs-text-track-display {
    opacity: 0
}
```

In any case, thank you for your time and sorry again for the delay in my answer.

**Comment by Kiro705** [comment]
Hi all,
I am reopening this issue because there is a PR which presents a potential solution. https://github.com/googleads/videojs-ima/pull/859

I am trying to gauge if this is a fix that will be utilized if it is implemented. If this issue does affect you and the solution looks like something you will implement, please reply here.

Thank you,
IMA SDK DevRel

**Comment by gcarrion-gfrmedia** [comment]
@Kiro705 Yes, please! We're also having this issue.

**Comment by Kiro705** [comment]
The pull request https://github.com/googleads/videojs-ima/pull/859 was merged and is live in version 1.7.5 of the plugin. If you would like to use it please upgrade to the new version.
